README.md
# GMail for Ruby

A Rubyesque interface to Google's GMail, with all the tools you'll need. Search, 
read and send multipart emails, archive, mark as read/unread, delete emails, 
and manage labels.

This is a fork from Kriss 'nu7hatch' Kowalik's gem that just adds 
batched envelope reads.

It's based on Daniel Parker's ruby-gmail gem. This version has more friendy
API, is well tested, better documented and have many other improvements.  

## Author(s)

* Jacob Carlson
* Kriss 'nu7hatch' Kowalik
* [Daniel Parker of BehindLogic.com](http://github.com/dcparker)
 
Extra thanks for specific feature contributions from:

* [abhishiv](http://github.com/abhishiv)
* [Michael Young](http://github.com/myoung8)
* [Nicolas Fouché](http://github.com/nfo)
* [Stefano Bernardi](http://github.com/stefanobernardi)
* [Benjamin Bock](http://github.com/bb)
* [Arthur Chiu](http://github.com/achiu)
* [Justin Perkins](http://github.com/justinperkins)
* [Mikkel Malmberg](http://github.com/mikker)
* [Julien Blanchard](http://github.com/julienXX)
* [Federico Galassi](http://github.com/fgalassi)

## Installation

You can install it easy using rubygems:

    sudo gem install gmail
    
Or install it manualy:

    git clone git://github.com/nu7hatch/gmail.git
    cd gmail
    rake install

To install gmail gem you have to met following requirements (with rubygems all 
will be installed automatically):

* mail
* mime
* smpt_tls (Ruby < 1.8.7)

## Features

* Search emails
* Read emails (handles attachments)
* Emails: label, archive, delete, mark as read/unread/spam, star
* Manage labels
* Create and send multipart email messages in plaintext and/or html, with inline 
  images and attachments
* Utilizes Gmail's IMAP & SMTP, MIME-type detection and parses and generates 
  MIME properly.

## Basic usage

First of all require the `gmail` library.

    require 'gmail'
    
### Authenticating gmail sessions

This will you automatically log in to your account. 

    gmail = Gmail.connect(username, password)
    # play with your gmail...
    gmail.logout

If you pass a block, the session will be passed into the block, and the session 
will be logged out after the block is executed.

    Gmail.connect(username, password) do |gmail|
      # play with your gmail...
    end
    
Examples above are "quiet", it means that it will not raise any errors when 
session couldn't be started (eg. because of connection error or invalid 
authorization data). You can use connection which handles errors raising:

    Gmail.connect!(username, password)
    Gmail.connect!(username, password) {|gmail| ... play with gmail ... }
    
You can also check if you are logged in at any time:

    Gmail.connect(username, password) do |gmail|
      gmail.logged_in?
    end

### XOAuth authentication

From v0.4.0 it's possible to authenticate with your Gmail account using XOAuth
method. It's very simple:

    gmail = Gmail.connect(:xoauth, "email@domain.com", 
      :token           => 'TOKEN',
      :secret          => 'TOKEN_SECRET',
      :consumer_key    => 'CONSUMER_KEY',
      :consumer_secret => 'CONSUMER_SECRET'
    )
    
For more information check out the [gmail_xoauth](https://github.com/nfo/gmail_xoauth)
gem from Nicolas Fouché.

### Counting and gathering emails
    
Get counts for messages in the inbox:

    gmail.inbox.count
    gmail.inbox.count(:unread)
    gmail.inbox.count(:read)

Count with some criteria:

    gmail.inbox.count(:after => Date.parse("2010-02-20"), :before => Date.parse("2010-03-20"))
    gmail.inbox.count(:on => Date.parse("2010-04-15"))
    gmail.inbox.count(:from => "myfriend@gmail.com")
    gmail.inbox.count(:to => "directlytome@gmail.com")

Combine flags and options:

    gmail.inbox.count(:unread, :from => "myboss@gmail.com")
    
Browsing labeled emails is similar to work with inbox.

    gmail.mailbox('Urgent').count
    
Getting messages works the same way as counting: Remember that every message in a 
conversation/thread will come as a separate message.

    gmail.inbox.emails(:unread, :before => Date.parse("2010-04-20"), :from => "myboss@gmail.com")

You can use also one of aliases:

    gmail.inbox.find(...)
    gmail.inbox.search(...)
    gmail.inbox.mails(...)    
    
Also you can manipulate each message using block style:

    gmail.inbox.find(:unread) do |email|
      email.read!
    end
    
### Working with emails!

Any news older than 4-20, mark as read and archive it:

    gmail.inbox.find(:before => Date.parse("2010-04-20"), :from => "news@nbcnews.com") do |email|
      email.read! # can also unread!, spam! or star!
      email.archive!
    end

Delete emails from X:

    gmail.inbox.find(:from => "x-fiance@gmail.com").each do |email|
      email.delete!
    end

Save all attachments in the "Faxes" label to a local folder:

    folder = "/where/ever"
    gmail.mailbox("Faxes").emails do |email|
      if !email.message.attachments.empty?
        email.message.save_attachments_to(folder)
      end
    end
     
You can use also `#label` method instead of `#mailbox`: 

    gmail.label("Faxes").emails {|email| ... }

Save just the first attachment from the newest unread email (assuming pdf):

    email = gmail.inbox.find(:unread).first
    email.attachments[0].save_to_file("/path/to/location")

Add a label to a message:

    email.label("Faxes")
    
Example above will raise error when you don't have the `Faxes` label. You can 
avoid this using:

    email.label!("Faxes") # The `Faxes` label will be automatically created now

You can also move message to a label/mailbox:
 
    email.move_to("Faxes")
    email.move_to!("NewLabel")
    
There is also few shortcuts to mark messages quickly:

    email.read!
    email.unread!
    email.spam!
    email.star!
    email.unstar!

### Managing labels

With Gmail gem you can also manage your labels. You can get list of defined 
labels:

    gmail.labels.all

Create new label:
  
    gmail.labels.new("Urgent")
    gmail.labels.add("AnotherOne")
    
Remove labels:

    gmail.labels.delete("Uregent")
    
Or check if given label exists:

    gmail.labels.exists?("Uregent") # => false
    gmail.labels.exists?("AnotherOne") # => true

### Composing and sending emails

Creating emails now uses the amazing [Mail](http://rubygems.org/gems/mail) rubygem. 
See its [documentation here](http://github.com/mikel/mail). The Ruby Gmail will 
automatically configure your Mail emails to be sent via your Gmail account's SMTP, 
so they will be in your Gmail's "Sent" folder. Also, no need to specify the "From" 
email either, because ruby-gmail will set it for you.

    gmail.deliver do
      to "email@example.com"
      subject "Having fun in Puerto Rico!"
      text_part do
        body "Text of plaintext message."
      end
      html_part do
        body "<p>Text of <em>html</em> message.</p>"
      end
      add_file "/path/to/some_image.jpg"
    end

Or, compose the message first and send it later

    email = gmail.compose do
      to "email@example.com"
      subject "Having fun in Puerto Rico!"
      body "Spent the day on the road..."
    end
    email.deliver! # or: gmail.deliver(email)

## Note on Patches/Pull Requests
 
* Fork the project.
* Make your feature addition or bug fix.
* Add tests for it. This is important so I don't break it in a
  future version unintentionally.
* Commit, do not mess with rakefile, version, or history.
  (if you want to have your own version, that is fine but bump version in a commit by itself I can ignore when I pull)
* Send me a pull request. Bonus points for topic branches.

## Copyright

* Copyrignt (c) 2010 Kriss 'nu7hatch' Kowalik
* Copyright (c) 2009-2010 BehindLogic

See LICENSE for details.



CHANGELOG.md
# Gmail gem changelog

## 0.4.5
* Added support for X-GM-MSGID and X-GM-THRID extensions (includes monkey path to Net::IMAP)
* Added support for Mailbox#unread to return just the envelope
* Mailbox#envelopes (and Mailbox#unread) now return an object called Envelope that's much easier to deal with than struct Net::IMAP::Envelope (and includes the X-GM-MSGID and X-GM-THRID attributes)

## 0.4.3
* Removed exception swallowing.

## 0.4.2
* Removed ore dependency that made installation more tedious than necessary.

## 0.4.1
* Added mailbox#envelopes to pull back the envelopes of all messages in a mailbox in 1 request.

## 0.4.0

* Added XOAUTH authentication method (Thanks Stefano Bernardi and Nicolas Fouché)
* Separated clients
* Fixed specs

## 0.3.4

* Fixes in mailbox filters shortcuts (Thanks Benjamin Bock)

## 0.3.3

* Added #expunge to Mailbox (Thanks Benjamin Bock)
* Added more mailbox filters (Thanks Benjamin Bock)
* Added shortcuts for mailbox filters
* Minor bugfixes

## 0.3.2

* Added envelope fetching
* Minor bugfixes

## 0.3.0

* Refactoring
* Fixed bugs
* API improvements
* Better documentation
* Code cleanup
* RSpec for everything

## 0.1.1 / 2010-05-11

* Added explicit tmail dependency in gemspec
* Added better README tutorial content

## 0.0.9 / 2010-04-17

* Fixed content-transfer-encoding when sending email

## 0.0.8 / 2009-12-23

* Fixed attaching a file to an empty message

## 0.0.7 / 2009-12-23

* Improved multipart message parsing reliability

## 0.0.6 / 2009-12-21

* Fixed multipart parsing for when the boundary is marked in quotes.

## 0.0.5 / 2009-12-16

* Fixed IMAP initializer to work with Ruby 1.9's net/imap
* Better logout depending on the IMAP connection itself
* Added MIME::Message#text and MIME::Message#html for easier access to an email body
* Improved the MIME-parsing API slightly
* Added some tests

## 0.0.4 / 2009-11-30

* Added label creation (thanks to Justin Perkins / http://github.com/justinperkins)
* Made the gem login automatically when first needed
* Added an optional block on the Gmail.new object that will login and logout for you
* Added several search options (thanks to Mikkel Malmberg / http://github.com/mikker)

## 0.0.3 / 2009-11-19

* Fixed MIME::Message#content= for messages without an encoding
* Added Gmail#new_message

## 0.0.2 / 2009-11-18

* Made all of the examples in the README possible

## 0.0.1 / 2009-11-18

* Birthday!



lib/gmail.rb
require 'net/imap'
require 'net/smtp'
require 'mail'
require 'date'
require 'time'

if RUBY_VERSION < "1.8.7"
  require "smtp_tls"
end

class Object
  def to_imap_date
    Date.parse(to_s).strftime("%d-%B-%Y")
  end
end

module Gmail
  autoload :Version, "gmail/version"
  autoload :Client,  "gmail/client"
  autoload :Labels,  "gmail/labels"
  autoload :Mailbox, "gmail/mailbox"
  autoload :Message, "gmail/message"

  class << self
    # Creates new Gmail connection using given authorization options.
    #
    # ==== Examples
    #
    #   Gmail.new(:plain, "foo@gmail.com", "password")
    #   Gmail.new(:xoauth, "foo@gmail.com", 
    #     :consumer_key => "",
    #     :consumer_secret => "",
    #     :token => "",
    #     :secret => "")
    #
    # To use plain authentication mehod you can also call:
    #
    #   Gmail.new("foo@gmail.com", "password")
    #
    # You can also use block-style call:
    #
    #   Gmail.new("foo@gmail.com", "password") do |client|
    #     # ...
    #   end
    #

    ['', '!'].each { |kind|
      define_method("new#{kind}") do |*args, &block|
        args.unshift(:plain) unless args.first.is_a?(Symbol)
        client = Gmail::Client.new(*args)
        client.send("connect#{kind}") and client.send("login#{kind}")
        if block_given?
          yield client
          client.logout
        end

        client
      end
    }

    alias :connect :new
    alias :connect! :new!
  end # << self

  # Wrapper class for the Net::IMAP structs that supports
  # the additional Gmail attributes like X-GM-MSGID and X-GM-THRID and
  # generally act more rubyish
  class Address
    attr_accessor :mailbox, :name, :route, :host

    def initialize(args = {})
      @mailbox = args[:src].mailbox
      @name = args[:src].name
      @route = args[:src].route
      @host = args[:src].host
    end

    def address 
      "#{@mailbox}@#{@host}"
    end
  end

  class Envelope
    attr_accessor :sender, :to, :reply_to, :cc, :bcc
    attr_accessor :subject, :date
    attr_accessor :in_reply_to, :message_id
    attr_accessor :gm_msg_id, :gm_thread_id

    def initialize(args={})
      @to = []
      @cc = []
      @bcc = []

      if args[:src]
        args[:src].to.to_a.each { |s| @to.push Address.new(:src => s) }
        args[:src].cc.to_a.each { |s| @cc.push Address.new(:src => s) }
        args[:src].bcc.to_a.each { |s| @bcc.push Address.new(:src => s) }
        @sender = Address.new(:src => args[:src].sender.first)
        @subject = args[:src].subject
        @date = args[:src].date
        @in_reply_to = args[:src].in_reply_to
        @message_id = args[:src].message_id
      end

      @gm_msg_id = args[:gm_msg_id]
      @gm_thread_id = args[:gm_thread_id]
    end

    def url
      return nil unless @gm_msg_id

      "https://mail.google.com/mail/#inbox/#{@gm_msg_id}"
    end
  end
end # Gmail

# Monkey patches to understand the extended Gmail atttributes, and add a
# url method to struct Net::IMAP::Envelope
module Net
  class IMAP
    class ResponseParser
      def msg_att
        match(T_LPAR)
        attr = {}
        while true
          token = lookahead
          case token.symbol
          when T_RPAR
            shift_token
            break
          when T_SPACE
            shift_token
            token = lookahead
          end
          case token.value
          when /\A(?:ENVELOPE)\z/ni
            name, val = envelope_data
          when /\A(?:FLAGS)\z/ni
            name, val = flags_data
          when /\A(?:INTERNALDATE)\z/ni
            name, val = internaldate_data
          when /\A(?:RFC822(?:\.HEADER|\.TEXT)?)\z/ni
            name, val = rfc822_text
          when /\A(?:RFC822\.SIZE)\z/ni
            name, val = rfc822_size
          when /\A(?:BODY(?:STRUCTURE)?)\z/ni
            name, val = body_data
          when /\A(?:UID)\z/ni
            name, val = uid_data
          when /\A(?:X-GM-MSGID)\z/ni
            name, val = uid_data
          when /\A(?:X-GM-THRID)\z/ni
            name, val = uid_data
          else
            parse_error("unknown attribute `%s'", token.value)
          end
          attr[name] = val
        end
        return attr
      end
    end # class ResponseParser
  end # class IMAP
end # module Net


lib/gmail/client/base.rb
require 'thread'

module Gmail
  module Client
    class Base
      # GMail IMAP defaults
      GMAIL_IMAP_HOST = 'imap.gmail.com'
      GMAIL_IMAP_PORT = 993
      
      # GMail SMTP defaults
      GMAIL_SMTP_HOST = "smtp.gmail.com"
      GMAIL_SMTP_PORT = 587
      
      attr_reader :username
      attr_reader :options
      
      def initialize(username, options={})
        defaults       = {}
        @username      = fill_username(username)
        @options       = defaults.merge(options)
        @mailbox_mutex = Mutex.new
      end
      
      # Connect to gmail service. 
      def connect(raise_errors = true)
        @imap = Net::IMAP.new(GMAIL_IMAP_HOST, GMAIL_IMAP_PORT, usessl = true,
                              certs = nil, verify = false)
      end
      
      # This version of connect will raise error on failure...
      def connect!
        connect(true)
      end
      
      # Return current connection. Log in automaticaly to specified account if 
      # it is necessary.
      def connection
        login and at_exit { logout } unless logged_in?
        @imap
      end
      alias :conn :connection
      
      # Login to specified account.
      def login(*args)
        raise NotImplementedError, "The `#{self.class.name}#login` method is not implemented."
      end
      alias :sign_in :login
      
      # This version of login will raise error on failure...
      def login!
        login(true)
      end
      alias :sign_in! :login!
      
      # Returns +true+ when you are logged in to specified account.
      def logged_in?
        !!@logged_in
      end
      alias :signed_in? :logged_in?
      
      # Logout from GMail service. 
      def logout
        @imap && logged_in? and @imap.logout
      ensure
        @logged_in = false
      end
      alias :sign_out :logout
      
      # Return labels object, which helps you with managing your GMail labels.
      # See <tt>Gmail::Labels</tt> for details.
      def labels
        @labels ||= Labels.new(conn)
      end
      
      # Compose new e-mail.
      #
      # ==== Examples
      #   
      #   mail = gmail.compose
      #   mail.from "test@gmail.org"
      #   mail.to "friend@gmail.com"
      #
      # ... or block style:
      #  
      #   mail = gmail.compose do 
      #     from "test@gmail.org"
      #     to "friend@gmail.com"
      #     subject "Hello!"
      #     body "Hello my friend! long time..."
      #   end
      #
      # Now you can deliver your mail:
      #
      #   gmail.deliver(mail)
      def compose(mail=nil, &block)
        if block_given?
          mail = Mail.new(&block)
        elsif !mail 
          mail = Mail.new
        end 

        mail.delivery_method(*smtp_settings)
        mail.from = username unless mail.from
        mail
      end
      alias :message :compose
      
      # Compose (optionaly) and send given email. 
      #
      # ==== Examples
      #
      #   gmail.deliver do 
      #     to "friend@gmail.com"
      #     subject "Hello friend!"
      #     body "Hi! How are you?"
      #   end
      #
      # ... or with already created message:
      #
      #   mail = Mail.new { ... }
      #   gmail.deliver(mail)
      #
      #   mail = gmail.compose { ... }
      #   gmail.deliver(mail)
      def deliver(mail=nil, &block)
        mail = compose(mail, &block) if block_given?
        mail.deliver!
      rescue Object => ex
        raise DeliveryError, "Couldn't deliver email: #{ex.to_s}"
      end
      
      # This version of deliver will raise error on failure...
      def deliver!(mail=nil, &block)
        deliver(mail, true, &block)
      end
      
      # Do something with given mailbox or within it context. 
      #
      # ==== Examples
      #
      #   mailbox = gmail.mailbox("INBOX")
      #   mailbox.emails(:all)
      #   mailbox.count(:unread, :before => Time.now-(20*24*3600))
      #
      # ... or block style:
      #
      #   gmail.label("Work") do |mailbox|
      #     mailbox.emails(:unread)
      #     mailbox.count(:all)
      #     ...
      #   end
      def mailbox(name, &block)
        @mailbox_mutex.synchronize do
          name = name.to_s
          mailbox = (mailboxes[name] ||= Mailbox.new(self, name))
          switch_to_mailbox(name) if @current_mailbox != name

          if block_given?
            mailbox_stack << @current_mailbox
            result = block.arity == 1 ? block.call(mailbox) : block.call
            mailbox_stack.pop
            switch_to_mailbox(mailbox_stack.last)
            return result
          end

          return mailbox
        end
      end
      alias :in_mailbox :mailbox
      alias :in_label :mailbox
      alias :label :mailbox
      
      # Alias for <tt>mailbox("INBOX")</tt>. See <tt>Gmail::Client#mailbox</tt>
      # for details.
      def inbox
        mailbox("INBOX")
      end
      
      def mailboxes
        @mailboxes ||= {}
      end
      
      def inspect
        "#<Gmail::Client#{'0x%04x' % (object_id << 1)} (#{username}) #{'dis' if !logged_in?}connected>"
      end
      
      def fill_username(username)
        username =~ /@/ ? username : "#{username}@gmail.com"
      end

      def mail_domain
        username.split('@')[0]
      end
      
      private
      
      def switch_to_mailbox(mailbox)
        if mailbox
          mailbox = Net::IMAP.encode_utf7(mailbox)
          conn.select(mailbox)
        end
        @current_mailbox = mailbox
      end
      
      def mailbox_stack
        @mailbox_stack ||= []
      end
      
      def smtp_settings
        [:smtp, {
          :address => GMAIL_SMTP_HOST,
          :port => GMAIL_SMTP_PORT,
          :domain => mail_domain,
          :user_name => username,
          :password => password,
          :authentication => 'plain',
          :enable_starttls_auto => true
        }]
      end
    end # Base
  end # Client
end # Gmail


lib/gmail/client/xoauth.rb
require 'gmail_xoauth'

module Gmail
  module Client
    class XOAuth < Base
      attr_reader :token
      attr_reader :secret
      attr_reader :consumer_key
      attr_reader :consumer_secret

      def initialize(username, options={})
        @token           = options.delete(:token)
        @secret          = options.delete(:secret)
        @consumer_key    = options.delete(:consumer_key)
        @consumer_secret = options.delete(:consumer_secret)
       
        super(username, options)
      end

      def login(raise_errors = true)
        @imap and @logged_in = (login = @imap.authenticate('XOAUTH', username,
          :consumer_key    => consumer_key,
          :consumer_secret => consumer_secret,
          :token           => token,
          :token_secret    => secret
        )) && login.name == 'OK'
      end

      def smtp_settings
        [:smtp, {
           :address => GMAIL_SMTP_HOST,
           :port => GMAIL_SMTP_PORT,
           :domain => mail_domain,
           :user_name => username,
           :password => {
             :consumer_key    => consumer_key,
             :consumer_secret => consumer_secret,
             :token           => token,
             :token_secret    => secret
           },
           :authentication => :xoauth,
           :enable_starttls_auto => true
         }]
      end
    end # XOAuth

    register :xoauth, XOAuth
  end # Client
end # Gmail


lib/gmail/mailbox.rb
module Gmail
  class Mailbox
    MAILBOX_ALIASES = {
      :all       => ['ALL'],
      :seen      => ['SEEN'],
      :unseen    => ['UNSEEN'],
      :read      => ['SEEN'],
      :unread    => ['UNSEEN'],
      :flagged   => ['FLAGGED'],
      :unflagged => ['UNFLAGGED'],
      :starred   => ['FLAGGED'],
      :unstarred => ['UNFLAGGED'], 
      :deleted   => ['DELETED'],
      :undeleted => ['UNDELETED'],
      :draft     => ['DRAFT'],
      :undrafted => ['UNDRAFT']
    }
  
    attr_reader :name
    attr_reader :external_name

    def initialize(gmail, name="INBOX")
      @name  = name
      @external_name = Net::IMAP.decode_utf7(name)
      @gmail = gmail
    end

    # Pull back the envelopes of all messages in a mailbox in an array.
    # This uses only 1 IMAP request, making it much faster than
    # using the email method.
    def envelopes(*args, &block)
      args << :all if args.size == 0

      if args.first.is_a?(Symbol) 
        search = MAILBOX_ALIASES[args.shift].dup
        opts = args.first.is_a?(Hash) ? args.first : {}
        
        opts[:after] and search.concat ['SINCE', opts[:after].to_imap_date]
        opts[:before] and search.concat ['BEFORE', opts[:before].to_imap_date]
        opts[:on] and search.concat ['ON', opts[:on].to_imap_date]
        opts[:from] and search.concat ['FROM', opts[:from]]
        opts[:to] and search.concat ['TO', opts[:to]]
        opts[:subject] and search.concat ['SUBJECT', opts[:subject]]
        opts[:label] and search.concat ['LABEL', opts[:label]]
        opts[:attachment] and search.concat ['HAS', 'attachment']
        opts[:search] and search.concat ['BODY', opts[:search]]
        opts[:body] and search.concat ['BODY', opts[:body]]
        opts[:query] and search.concat opts[:query]

        @gmail.mailbox(name) do
          msg_ids = @gmail.conn.uid_search(search)
          return [] if msg_ids.empty?
          attrs = ["ENVELOPE", "X-GM-MSGID", "X-GM-THRID"]
          return @gmail.conn.uid_fetch(msg_ids, attrs).collect do |data|
            Envelope.new(:src => data.attr['ENVELOPE'],
                         :gm_msg_id => data.attr['X-GM-MSGID'],
                         :gm_thread_id => data.attr['X-GM-THRID'])
          end
        end
      elsif args.first.is_a?(Hash)
        emails(:all, args.first)
      else
        raise ArgumentError, "Invalid search criteria"
      end
    end

    # Return the envelopes of all unread messages in the mailbox.
    def unread(*args, &block)
      @gmail.mailbox(name) do
        msg_ids = @gmail.conn.uid_search('UNSEEN')
        return [] if msg_ids.empty?

        attrs = ["ENVELOPE", "X-GM-MSGID", "X-GM-THRID"]
        return @gmail.conn.uid_fetch(msg_ids, attrs).collect do |data|
          Envelope.new(:src => data.attr['ENVELOPE'],
                       :gm_msg_id => data.attr['X-GM-MSGID'],
                       :gm_thread_id => data.attr['X-GM-THRID'])
        end
      end
    end

    # Returns list of emails which meets given criteria. 
    #
    # ==== Examples
    #
    #   gmail.inbox.emails(:all)
    #   gmail.inbox.emails(:unread, :from => "friend@gmail.com")
    #   gmail.inbox.emails(:all, :after => Time.now-(20*24*3600))
    #   gmail.mailbox("Test").emails(:read)
    #
    #   gmail.mailbox("Test") do |box| 
    #     box.emails(:read)
    #     box.emails(:unread) do |email|
    #       ... do something with each email...
    #     end
    #   end
    def emails(*args, &block)
      args << :all if args.size == 0

      if args.first.is_a?(Symbol) 
        search = MAILBOX_ALIASES[args.shift].dup
        opts = args.first.is_a?(Hash) ? args.first : {}
        
        opts[:after] and search.concat ['SINCE', opts[:after].to_imap_date]
        opts[:before] and search.concat ['BEFORE', opts[:before].to_imap_date]
        opts[:on] and search.concat ['ON', opts[:on].to_imap_date]
        opts[:from] and search.concat ['FROM', opts[:from]]
        opts[:to] and search.concat ['TO', opts[:to]]
        opts[:subject] and search.concat ['SUBJECT', opts[:subject]]
        opts[:label] and search.concat ['LABEL', opts[:label]]
        opts[:attachment] and search.concat ['HAS', 'attachment']
        opts[:search] and search.concat ['BODY', opts[:search]]
        opts[:body] and search.concat ['BODY', opts[:body]]
        opts[:query] and search.concat opts[:query]

        @gmail.mailbox(name) do
          @gmail.conn.uid_search(search).collect do |uid| 
            message = (messages[uid] ||= Message.new(self, uid))
            block.call(message) if block_given?
            message
          end
        end
      elsif args.first.is_a?(Hash)
        emails(:all, args.first)
      else
        raise ArgumentError, "Invalid search criteria"
      end
    end
    alias :mails :emails
    alias :search :emails
    alias :find :emails
    alias :filter :emails

    # This is a convenience method that really probably shouldn't need to
    # exist, but it does make code more readable, if seriously all you want
    # is the count of messages.
    #
    # ==== Examples
    #
    #   gmail.inbox.count(:all)
    #   gmail.inbox.count(:unread, :from => "friend@gmail.com")
    #   gmail.mailbox("Test").count(:all, :after => Time.now-(20*24*3600))
    def count(*args)
      emails(*args).size
    end

    # This permanently removes messages which are marked as deleted
    def expunge
      @gmail.mailbox(name) { @gmail.conn.expunge }
    end

    # Cached messages. 
    def messages
      @messages ||= {}
    end
    
    def inspect
      "#<Gmail::Mailbox#{'0x%04x' % (object_id << 1)} name=#{external_name}>"
    end

    def to_s
      name
    end

    # Disabled 07.28.2011 because we want return just the envelopes,
    # not the complete messages.
=begin
    MAILBOX_ALIASES.each_key { |mailbox|
      define_method(mailbox) do |*args, &block|
        emails(mailbox, *args, &block)
      end
    }
=end
  end # Message
end # Gmail
